os/kernel/binary_manager/binary_manager_recovery.c

```diff
@@ -0,0 +1,204 @@
+	if (tcb == NULL || tcb->group == NULL || tcb->group->tg_loadtask < 0) {
+		bmdbg("Failed to get pid %d binary info\n", pid);
+		return ERROR;
+	}
+
+	bmvdbg("Faulty id %d, binary id %d\n", pid, tcb->group->tg_loadtask);
+
+	return tcb->group->tg_loadtask;
+}
+
+/****************************************************************************
+ * Name: binary_manager_kill_binary
+ *
+ * Description:
+ *   This function will kill all the task/thread created by the binary
+ *   i.e input pid.
+ *   If any fault occurs in the system, recovery thread will call this function to
+ *   kill all the thread/task created by the input binary.
+ *
+ * Input parameters:
+ *   pid   -   The pid of the binary, whoes all childs to be killed
+ *
+ * Returned Value:
+ *   Zero (OK) on success; otherwise -1 (ERROR) value is returned.
+ *
+ ****************************************************************************/
+static int binary_manager_kill_binary(pid_t pid, int binid)
+{
+	int ret;
+
+	binmgr_kill_data_t info;
+
+	if (pid < 0 || binid < 0) {
+		return ERROR;
+	}
+
+	/* Set binary id and faulty id */
+	info.binid = binid;
+	info.faultid = pid;
+
+	/* Search all task/pthreads created by same loaded task */
+	sched_foreach(recovery_manager_kill_each, (FAR void *)&info);
+
+	/* Finally, unload binary */
+	ret = task_terminate(binid, true);
+	if (ret < 0) {
+		bmdbg("Failed to unload binary %d, ret %d, errno %d\n", binid, ret, errno);
+		return ERROR;
+	}
+	bmvdbg("Unload binary! pid %d\n", binid);
+
+	return OK;
+}
+
+/****************************************************************************
+ * Name: recovery_thread
+ *
+ * Description:
+ *   This function will receive the faulty binary's pid and check if it is one
+ *   of the registered binary with binary manager.
+ *   If the binary is registered then it will be restarted after killing its
+ *   child processes, and if it is not registered then board will be rebooted.
+ *
+ * Input parameters:
+ *   None
+ *
+ * Returned Value:
+ *   Zero (OK) on success; otherwise -1 (ERROR) value is returned.
+ *
+ * Assumptions:
+ *    All the binaries of the system will be registered with the binary manager
+ *
+ ****************************************************************************/
+int recovery_thread(int argc, char *argv[])
+{
+	int ret;
+	int binidx;
+	int binid;
+	pid_t pid;
+
+	pid = (pid_t)atoi(argv[1]);
+
+	bmvdbg("Faulty pid is %d\n", pid);
+
+	/* Get binary id of fault task and check it is registered in binary manager */
+	binid = binary_manager_get_binid(pid);
+	if (binid < 0) {
+		bmdbg("Invliad value, pid %d, binid %d\n", pid, binid);
+		goto reboot_board;
+	}
+
+	binidx = binary_manager_get_index_with_binid(binid);
+	if (binidx < 0) {
+		bmdbg("binary pid %d is not registered to binary manager\n", binid);
+		goto reboot_board;
+	}
+
+	/* Kill its children and restart binary if the binary is registered with the binary manager */
+	ret = binary_manager_kill_binary(pid, binid);
```

Comment:
Where is the notice to another binary about dead of this binary?

Reply:
That implementation is not finished yet. I will update.